src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -3105,7 +3105,10 @@ internal override int CompareKeys(int index1, int index2)
                 if (next == null) return index1 - index2;
                 return next.CompareKeys(index1, index2);
             }
-            return descending ? -c : c;
+            // -c will result in a negative value for int.MaxValue (-int.MinValue == int.MinValue).
+            // Flipping keys earlier is more likely to trigger something strange in a comparer,
+            // particularly as it comes to the sort being stable.
+            return (descending ^ (c > 0)) ? 1 : -1;
```

Comment:
:exclamation: Use `!=` instead of `^`.

:memo: Some people understand `^` for Boolean values. Everyone understands `!=` for Boolean values. Functionally they are equivalent.